### ISSUE TITLE
Bad Contrast with waveshare board

### ISSUE THREAD

**Issue by jhughesoti** [wuspy/portal_calendar#18]
I initially set this up with an esp32 I had on hand but it only worked once then would not update the display anymore.  I bought this board https://www.waveshare.com/wiki/E-Paper_ESP32_Driver_Board#Pin to remove doubt about the board and after tweaking the code a bit to make the spi pins compatible, it does update but the image is super faint.  I can run the waveshare demos and do not see this faint image and text so I know the display is fine as well as the connection.  Only difference is using this library vs theirs.  Any thoughts on what could be the difference in causing the bad contrast on the screen?

**Comment by wuspy** [comment]
Does the display look similar to #4? The display code this uses is different from the waveshare examples because it supports 4-color grayscale, with the downside that it may not work well on every display. Unfortunately there's not much I can do to help since I'm no expert on e-paper tech, and both of the displays I have work fine so I can't test out different settings. There are a few things you can try though.

1. Adjust the VCOM_DC value in DisplayGDEW075T7.cpp. That will definitely change the contrast, I've thought about moving it to config.h under a contrast setting but I don't know if adjusting it too far one way or the other could damage the display over time.
    ```cpp
    sendCommand(CMD_VDCS);  // VCOM DC Setting (min 0x00 = -0.1v, max 0x4F = -4.05v)
    sendData(0x26);         // -2.0v
    ```
2. Try out the stock_waveform_test branch, that uses the same settings as the waveshare example but grayscale won't work. It looks like the only non-grayscale related settings I changed were the values for VCOM_DC and CDI. This project could definitely be made to work with the stock settings, the graphics would just need to be changed to replace grayscale with dithering.
3. If all else fails return your display and buy another one, if you're able to. I don't know the percentage of display that have issues with these settings but you're only the 2nd person to open an issue about this so hopefully it's not that high.

**Comment by jhughesoti** [comment]
Yes, that is the exact scenario I was having.  I attempted to adjust the VCOM setting all the way up to 0x35 incrementally and saw very little improvement.  I stopped at 0x35 due to the note concerning possible damage to the screen based on this setting so wanted to stay away from the extremes.  Some improvements in beginning of the display but not much if anything towards the middle and bottom. 

 I switched to the stock_waveform and that does look a lot better but since this has no grayscale, any icon at the bottom that was supposed to be faded out, was just gone.  Not the end of the world if I add the weather since I think there's nothing left to need the grayscale but not 100% on that one.

I have also opened a ticket with waveshare and shared the dim image so we'll see what they say.

**Comment by wuspy** [comment]
I pushed another branch called alternate_display_settings_test that uses the same settings as GxEPD2_4G if you wanna try that out. The current settings are a mix between that project's and waveshare's.

**Comment by jhughesoti** [comment]
Just gave that branch a shot but looks very much the same as the main branch.  Tried to modify the VCOM_DC value just to see what would happen but no luck there either.

**Comment by jhughesoti** [comment]
For what it's worth, I implemented a weather api key and am running on the stock_waveform branch so it looks good.  I can switch back to the icons with the boot button easily if there's anything to test further.

**Comment by jhughesoti** [comment]
I received replacement hardware from waveshare and confirmed that I had a bad hat which was making connections to the wire harness impossible.  After confirming this, I ran through a series of tests to check other scenario's where I had seen issues with the image.  I will leave the results here for anybody else finding similar issues:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/JONATH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/JONATH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



e-ink esp32 dev   board | Main Branch | stock_waveform_test | alternate_display_settings_test
-- | -- | -- | --
Built in ribbon connection | faint overall image, greyscale icons basically invisible | Good dark images, no greyscale, faded icons missing entirely | Good dark images, no greyscale, faded icons missing entirely
Wire harness to driver hat | All Good | Good dark images, no greyscale, faded icons missing entirely | Good dark images, no greyscale, faded icons missing entirely



</body>

</html>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/JONATH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/JONATH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Node MCU 32s | Main Branch | stock_waveform_test | alternate_display_settings_test
-- | -- | -- | --
Wire harness to driver hat | All good | Good dark images, no greyscale, faded icons missing entirely | Good dark images, no greyscale, faded icons missing entirely



</body>

</html>

In summary, from my testing, I wouldn't use the e-ink esp32 with the built in ribbon connection unless you only want to use the weather option and prefer simpler wiring but don't use the main branch if you do or the image will not be good.  Main branch is the only one that actually worked with grey scale for me so any generic esp32 should work and if there are issues with the image, it's probably due to a faulty hat rather than the screen.
